text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,399 @@
+# Meta
+[meta]: #meta
+- Name: Offline Buildpackages
+- Start Date: 2020-04-27
+- Author(s): dwillist
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: NA
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a method to attach dependencies to buildpackages and make these dependencies available during buildpack execution.
+
+# Motivation
+[motivation]: #motivation
+Why should we do this?
+    
+    1) Registry space optimizations.  
+    2) 1 to 1 correspondence between id+version and buildpack bits for buildpack authors.
+
+# What it is
+[what-it-is]: #what-it-is
+
+### Define the target persona:
+Buildpack user, Platform operator, Buildpack author
+
+### Explaining the feature largely in terms of examples.
+
+Let's take a quick look at how offline builds currently work.
+
+Say we are building an app that requires `ruby`. To perform an offline build a buildpack needs to:
+1) include a local copy of the `ruby` dependency.
+2) implement logic to copy the local `ruby` dependency into to a [`<layer>`](https://github.com/buildpacks/spec/blob/main/buildpack.md#layer-types).
+
+An offline buildpack would have a directory structure similar to this:
+
+```
+ruby_buildpack
+    ├── bin
+    │   ├── build
+    │   └── detect
+    ├── buildpack_dependencies (only present in offline buildpack)
+    │   ├── ruby-v2.7.1
+    │   └── ruby-v2.6.8
+    └── buildpack.toml
+```
+
+The above structure has some undesirable consequences:
+
+- offline and online buildpacks have different contents and therefore must have different buildpack `id` fields.
+
+We propose an extension to buildpackages that allows buildpack dependencies to be moved out of the buildpack archive and into a layer in the buildpackage. As well as, a lifecycle mechanism for letting buildpacks access these dependencies.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Lets walk through the following steps:
+1) How to create an offline buildpackage.
+2) Changes to the buildpackage format.
+3) How buildpacks access offline dependencies during execution.
+
+
+## How to Create an offline buildpackage
+This is currently done via the `pack package-buildpack` command, which consumes a [`package.toml`](https://buildpacks.io/docs/reference/package-config/) file. We propose adding an `offline_dependencies` array. We require that the `id` + `version` combination will be unique among all `offline_dependencies` entries.
```

Comment:
Does the version always have to be exact? Or can this support a range? Also, would version be optional?

Reply:
It would totally be up to the buildpack author what to put in the `version` field. It could really be any string. This is just a piece metadata telling your buildpack where it should look for this dependency in case of an offline build `/cnb/deps/<id>/<version>`. The only request is that it be unique.